### PR TITLE
refactor: refactor trace in redis & sql & mongo

### PR DESCRIPTION
**Pull request by chenquan**
Signed-off-by: chenquan <chenquan.dev@gmail.com>

### CODE REVIEW

**Comment by chenquan** [comment]
@kevwan PTAL

File: core/stores/mon/model.go [hunk 1]
```diff
@@ -169,31 +169,45 @@ func (m *Model) FindOneAndUpdate(ctx context.Context, v, filter interface{}, upd
...
-	ctx, span := startSpan(ctx)
-	defer span.End()
+// AbortTransaction implements the mongo.Session interface.
+func (w *wrappedSession) AbortTransaction(ctx context.Context) (err error) {
+	ctx, span := startSpan(ctx, "AbortTransaction")
+	defer func() {
+		endSpan(span, err)
+	}()
 
-	return w.brk.DoWithAcceptable(func() error {
```
Comment (kevwan) [inline core/stores/mon/model.go:176]: Use `return` directly works fine.

File: core/stores/sqlx/tx.go [hunk 2]
```diff
@@ -101,13 +113,16 @@ func (t txSession) QueryRowsPartial(v interface{}, q string, args ...interface{}
...
+	args ...interface{}) (err error) {
+	ctx, span := startSpan(ctx, q)
+	defer func() {
+		endSpan(span, err)
+	}()
 
 	return query(ctx, t.Tx, func(rows *sql.Rows) error {
 		return unmarshalRows(v, rows, false)
 	}, q, args...)
+
```
Comment (zhoushuguang) [inline core/stores/sqlx/tx.go:125]: Excess blank lines

File: core/stores/sqlx/tx.go [hunk 3]
```diff
@@ -101,13 +113,16 @@ func (t txSession) QueryRowsPartial(v interface{}, q string, args ...interface{}
 }
 
 func (t txSession) QueryRowsPartialCtx(ctx context.Context, v interface{}, q string,
-	args ...interface{}) error {
-	ctx, span := startSpan(ctx)
-	defer span.End()
+	args ...interface{}) (err error) {
```
Comment (zhoushuguang) [inline core/stores/sqlx/tx.go:116]: (err error) to error?

File: core/stores/sqlx/tx.go [hunk 4]
```diff
@@ -30,20 +30,26 @@ func (t txSession) Exec(q string, args ...interface{}) (sql.Result, error) {
 	return t.ExecCtx(context.Background(), q, args...)
 }
 
-func (t txSession) ExecCtx(ctx context.Context, q string, args ...interface{}) (sql.Result, error) {
-	ctx, span := startSpan(ctx)
-	defer span.End()
+func (t txSession) ExecCtx(ctx context.Context, q string, args ...interface{}) (result sql.Result, err error) {
+	ctx, span := startSpan(ctx, q)
+	defer func() {
```
Comment (zhoushuguang) [inline core/stores/sqlx/tx.go:35]: why not `defer endSpan(span, err)`？

File: core/stores/sqlx/sqlconn.go [hunk 5]
```diff
@@ -303,9 +316,11 @@ func (s statement) Exec(args ...interface{}) (sql.Result, error) {
 	return s.ExecCtx(context.Background(), args...)
 }
 
-func (s statement) ExecCtx(ctx context.Context, args ...interface{}) (sql.Result, error) {
-	ctx, span := startSpan(ctx)
-	defer span.End()
+func (s statement) ExecCtx(ctx context.Context, args ...interface{}) (result sql.Result, err error) {
```
Comment (zhoushuguang) [inline core/stores/sqlx/sqlconn.go:319]: it be better to return anonymous values？result and err are not used

File: core/stores/mon/model.go [hunk 6]
```diff
@@ -169,45 +179,79 @@ func (m *Model) FindOneAndUpdate(ctx context.Context, v, filter interface{}, upd
...
 
 	return w.brk.DoWithAcceptable(func() error {
+		starTime := timex.Now()
+		defer func() {
+			logDuration(ctx, w.name, commitTransaction, starTime, err)
+		}()
+
 		return w.Session.CommitTransaction(ctx)
 	}, acceptable)
+
```
Comment (zhoushuguang) [inline core/stores/mon/model.go:214]: blank line

File: core/stores/sqlx/tx.go (hunk 4 above)
Comment (chenquan) [reply core/stores/sqlx/tx.go:35]: defer call depends on the err value. the err might be changed later.

File: core/stores/sqlx/sqlconn.go (hunk 5 above)
Comment (chenquan) [reply core/stores/sqlx/sqlconn.go:319]: he defer call depends on the err value. the err might be changed later.

File: core/stores/sqlx/tx.go (hunk 3 above)
Comment (chenquan) [reply core/stores/sqlx/tx.go:116]: we need err, which is used when defer calls.

File: core/stores/mon/model.go (hunk 6 above)
Comment (chenquan) [reply core/stores/mon/model.go]: fixed

File: core/stores/sqlx/tx.go (hunk 2 above)
Comment (chenquan) [reply core/stores/sqlx/tx.go]: fixed

File: core/stores/redis/hook.go [hunk 7]
```diff
@@ -54,17 +58,22 @@ func (h hook) AfterProcess(ctx context.Context, cmd red.Cmder) error {
...
 }
 
 func (h hook) AfterProcessPipeline(ctx context.Context, cmds []red.Cmder) error {
-	h.endSpan(ctx)
-
 	if len(cmds) == 0 {
 		return nil
 	}
 
+	err := cmds[0].Err()
```
Comment (kevwan) [inline core/stores/redis/hook.go:74]: only the first err is processed?

File: core/stores/sqlx/sqlconn.go [hunk 8]
```diff
@@ -140,8 +137,10 @@ func (db *commonSqlConn) Exec(q string, args ...interface{}) (result sql.Result,
 
 func (db *commonSqlConn) ExecCtx(ctx context.Context, q string, args ...interface{}) (
 	result sql.Result, err error) {
-	ctx, span := startSpan(ctx)
-	defer span.End()
+	ctx, span := startSpan(ctx, q)
```
Comment (kevwan) [inline core/stores/sqlx/sqlconn.go:140]: span name might be too long.

File: core/stores/redis/hook.go (hunk 7 above)
Comment (chenquan) [reply core/stores/redis/hook.go:74]: https://github.com/go-redis/redis/blob/fa515240d4a060c0fe72e7c41f1416627ca3e29f/redis.go#L107-L110

all cmds have the same err.

File: core/stores/sqlx/sqlconn.go (hunk 8 above)
Comment (chenquan) [reply core/stores/sqlx/sqlconn.go]: done.

File: core/stores/redis/hook.go (hunk 7 above)
Comment (chenquan) [reply core/stores/redis/hook.go]: done

**Review by kevwan — APPROVED** [review]
LGTM.

**Pull request merged**